### PR TITLE
Add azure blob storage backend

### PR DESCRIPTION
**Pull request by eliecharra**
| Q                 | A
| ----------------- | ---
| 🐛 Bug fix?       | no
| 🚀 New feature?   | yes
| ⚠ Deprecations?   | no
| ❌ BC Break       | no
| 🔗 Related issues | #352 
| ❓ Documentation  | https://github.com/snyk/driftctl-docs/pull/222

## Description

ACC test pass [here](https://app.circleci.com/pipelines/github/snyk/driftctl/4243/workflows/4cfbfd0e-931c-47b0-961c-cacabf4f41c5/jobs/9248)

### CODE REVIEW

File: pkg/helpers/azure/blob.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package azure
```
Comment (eliecharra) [inline pkg/helpers/azure/blob.go:1]: I was not sure where to put that (new `helpers/azure` helpers), azure SDK is very weird.
We cannot reuse default auth mechanism than the one used by the `az` CLI, because it seems that this logic is done in the CLI itself not in the SDK.

I really don't like the `helper` pkg approach, it often became an big hold-all mess and prevent us to do a really good separation of concerns.

Any thoughts ?

File: pkg/iac/terraform/state/backend/azureblob_reader.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	"github.com/pkg/errors"
+	"github.com/snyk/driftctl/pkg/helpers/azure"
+)
+
+const BackendKeyAzureBlob = "azureblob"
```
Comment (eliecharra) [inline pkg/iac/terraform/state/backend/azureblob_reader.go:14]: Not sure if I should follow [terraform naming](https://www.terraform.io/language/settings/backends/azurerm) here, `azurerm` does not resonates to me.

WDYT ?

File: pkg/analyser/analysis.go [hunk 3]
```diff
@@ -128,10 +128,22 @@ func (a *Analysis) UnmarshalJSON(bytes []byte) error {
...
+			// TerraformStateSource.
+			// TODO: Add a discriminator field to be able to serialize back to the right type
+			// when we'll introduce a new source type
+			res.Source = &resource.TerraformStateSource{
+				State:  m.Source.S,
+				Module: m.Source.Ns,
+				Name:   m.Source.Name,
+			}
+		}
+		a.AddManaged(res)
```
Comment (eliecharra) [inline pkg/analyser/analysis.go:146]: This PR can live without that (test assertions will be less accurate)

Happy to do that in another PR if needed, and sorry for the noise for reviewers.

File: pkg/iac/terraform/state/backend/azureblob_reader.go (hunk 2 above)
Comment (wbeuil) [reply pkg/iac/terraform/state/backend/azureblob_reader.go:14]: For users that use this kind of backend, it would make more sense use the same naming.

Comment (eliecharra) [reply pkg/iac/terraform/state/backend/azureblob_reader.go]: Renamed ✅

**Review by wbeuil — CHANGES_REQUESTED** [review]

File: pkg/iac/terraform/state/enumerator/state_enumerator.go [hunk 4]
```diff
@@ -11,18 +11,20 @@ type StateEnumerator interface {
 	Enumerate() ([]string, error)
 }
 
-func GetEnumerator(config config.SupplierConfig) StateEnumerator {
+func GetEnumerator(config config.SupplierConfig) (StateEnumerator, error) {
 
 	switch config.Backend {
 	case backend.BackendKeyFile:
-		return NewFileEnumerator(config)
+		return NewFileEnumerator(config), nil
```
[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/iac/terraform/state/enumerator/state_enumerator.go:18]: I asked myself about it, but I wanted to keep that PR scope as small as possible.
I think we can always update those methods to return error the day we'll need it.

It does not add a lot of inconsistency IMO

File: pkg/helpers/azure/blob.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+	"github.com/pkg/errors"
+)
+
+func GetBlobSharedKey() (*azblob.SharedKeyCredential, error) {
+	storageAccountName, exist := os.LookupEnv("AZURE_STORAGE_ACCOUNT")
+	if !exist {
+		return nil, errors.New("AZURE_STORAGE_ACCOUNT should be defined to be able to read state from azure backend")
+	}
+
+	storageAccountKey, exist := os.LookupEnv("AZURE_STORAGE_KEY")
```
[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/helpers/azure/blob.go:16]: I wanted to follow the [azure standard](https://docs.microsoft.com/en-us/cli/azure/storage/account?view=azure-cli-latest#az-storage-account-generate-sas-optional-parameters) 

Do you think I should add those flags (`--account-key` and `--account-name`) in addition of env vars ?

[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/helpers/azure/blob.go:16]: Maybe we'll need to namespace them `--azure-account-key` and `--azure-account-name`, but they will no be standard anymore.

WDYT @sjourdan ?

[the remark this answers is not in the document]
Comment (wbeuil) [reply pkg/helpers/azure/blob.go:16]: Yes that's what I have in mind actually, 2 flags that we could reuse as `DCTL_AZURE_...` just like we did for the terraform cloud one.

[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/helpers/azure/blob.go:16]: Do you think we should prefix azure env vars with `DCTL_` ? 
I hate when I can't reuse SDK default configuration in third party tooling personally

[the remark this answers is not in the document]
Comment (wbeuil) [reply pkg/helpers/azure/blob.go:16]: The `DCTL_` prefix will come from our flags (e.g. `bindEnvToFlags`). It's only for the container storage account and storage key I meant

[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/helpers/azure/blob.go:16]: I think we should bind default azure env var to those flags, it's really annoying when you have to duplicate your env vars just because tooling is not using the same conventions.

I'm not against the flags, even if I don't find them really useful since we can use env vars. 
But at least they should be bound to default azure env vars `AZURE_STORAGE_ACCOUNT` and `AZURE_STORAGE_KEY` IMO, let me try to do that.

[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/helpers/azure/blob.go]: I changed the approach, and removed `blob.go` helper function, WDYT of that [b1d8d0a](https://github.com/snyk/driftctl/pull/1357/commits/b1d8d0a5c614fc0ef715371858b8461a04618c2e) @wbeuil ?

File: pkg/helpers/azure/blob.go (hunk 1 above)
Comment (eliecharra) [reply pkg/helpers/azure/blob.go]: That file does not makes sense since we now use flags

File: test/acceptance/testing.go [hunk 6]
```diff
@@ -115,7 +116,7 @@ func (c *AccTestCase) validate() error {
 		return fmt.Errorf("checks attribute must be defined")
 	}
 
-	if len(c.Paths) < 1 {
+	if len(c.Paths) < 1 && !c.DoNotRunTerraform {
```
Comment (wbeuil) [inline test/acceptance/testing.go:119]: What was the problem here?

File: pkg/helpers/azure/blob.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (wbeuil) [reply pkg/helpers/azure/blob.go]: I'm good with that approach so that we don't have `DCTL_*` env variable working and we can have both your 2 envs and flags working as expected!

**Review by wbeuil — APPROVED** [review]
LGTM, good job @eliecharra

File: pkg/helpers/azure/blob.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/helpers/azure/blob.go]: > I'm good with that approach so that we don't have DCTL_* env variable working and we can have both your 2 envs and flags working as expected!

To be honest I didn't tested `DCTL_` flags but I think they may work too.

[the remark this answers is not in the document]
Comment (wbeuil) [reply pkg/helpers/azure/blob.go]: I tested and it does not work with that:

```
if opts.StorageKey == "" || opts.StorageAccount == "" {
		return nil, errors.New("AZURE_STORAGE_ACCOUNT and AZURE_STORAGE_KEY should be defined to be able to read state from azure backend")
	}
```
